poc_mobile_verifier/src/speedtests.rs

```diff
@@ -0,0 +1,551 @@
+use crate::{Error, Result};
+use chrono::{DateTime, Duration, NaiveDateTime, Utc};
+use file_store::{
+    file_sink,
+    speedtest::{CellSpeedtest, CellSpeedtestIngestReport},
+    traits::{MsgDecode, TimestampEncode},
+    FileStore, FileType,
+};
+use futures::stream::{self, StreamExt, TryStreamExt};
+use helium_crypto::PublicKey;
+use helium_proto::services::poc_mobile as proto;
+use sqlx::{
+    postgres::{types::PgHasArrayType, PgTypeInfo},
+    FromRow, Type,
+};
+use std::{
+    collections::{HashMap, VecDeque},
+    ops::Range,
+};
+
+#[derive(Debug, Clone, Type)]
+#[sqlx(type_name = "speedtest")]
+pub struct Speedtest {
+    pub timestamp: NaiveDateTime,
+    pub upload_speed: i64,
+    pub download_speed: i64,
+    pub latency: i32,
+}
+
+impl Speedtest {
+    pub fn new(
+        timestamp: NaiveDateTime,
+        upload_speed: i64,
+        download_speed: i64,
+        latency: i32,
+    ) -> Self {
+        Self {
+            timestamp,
+            upload_speed,
+            download_speed,
+            latency,
+        }
+    }
+}
+
+impl PgHasArrayType for Speedtest {
+    fn array_type_info() -> PgTypeInfo {
+        PgTypeInfo::with_name("_speedtest")
+    }
+}
+
+#[derive(FromRow)]
+pub struct SpeedtestRollingAverage {
+    pub id: PublicKey,
+    pub speedtests: Vec<Speedtest>,
+    pub latest_timestamp: NaiveDateTime,
+}
+
+impl SpeedtestRollingAverage {
+    pub fn new(id: PublicKey) -> Self {
+        Self {
+            id,
+            speedtests: Vec::new(),
+            latest_timestamp: NaiveDateTime::default(),
+        }
+    }
+
+    pub async fn save(self, exec: impl sqlx::PgExecutor<'_>) -> Result<bool> {
+        #[derive(FromRow)]
+        struct SaveResult {
+            inserted: bool,
+        }
+
+        sqlx::query_as::<_, SaveResult>(
+            r#"
+            insert into speedtests (id, speedtests, latest_timestamp)
+            values ($1, $2, $3)
+            on conflict (id) do update set
+            speedtests = EXCLUDED.speedtests
```

Comment:
don't you need to update the latest_timestamp on conflict?

Reply:
yup